### PR TITLE
Aj/missing map type

### PR DESCRIPTION
**Pull request by veriAjinkya**
* Fixed bug in string replacement
* Fixed search for gep while reading and writing to structs

### CODE REVIEW

File: lib/libyul2llvm/YulASTVisitor/IntrinsicEmitter.cpp [hunk 1]
```diff
@@ -572,8 +572,6 @@ void YulIntrinsicHelper::handleCopyStructToStorage(YulFunctionCallNode &node) {
     std::abort();
   }
   if (!srcPtr->getType()->isPointerTy() ||
-      srcPtr->getType()->getPointerAddressSpace() !=
-          static_cast<int>(YulAddressSpace::MEMORY) ||
       !srcPtr->getType()->getPointerElementType()->isStructTy()) {
     llvm::errs() << "Source pointer not a pointer to memory struct";
```
Comment (Technius) [inline lib/libyul2llvm/YulASTVisitor/IntrinsicEmitter.cpp:576]: nit:
```suggestion
    llvm::errs() << "Source pointer not a pointer to a struct";
```

File: lib/libyul2llvm/YulASTVisitor/IntrinsicRewriter.cpp [hunk 2]
```diff
@@ -462,6 +462,7 @@ llvm::Value *YulIntrinsicHelper::getPointerToMemoryStructFieldFromRefrence(
     std::vector<unsigned int> &distances) {
   std::string instName = inst->getOperand(0)->getName().str();
   if (!inst->getPointerOperandType()->getPointerElementType()->isStructTy()) {
+    inst->dump();
```
Comment (Technius) [inline lib/libyul2llvm/YulASTVisitor/IntrinsicRewriter.cpp:465]: This needs to be removed or guarded by an `#ifndef _NDEBUG`

File: lib/libYulAST/Types/YulTypeContext.cpp [hunk 3]
```diff
@@ -682,9 +682,17 @@ YulTypeContext::convertStorageLayoutTypeNameToYulTypeName(std::string name) {
...
-  name = replace(name, "_$_", ",");
-  name = replace(name, "$_", "(");
+  // In when yul names are converted to storage layout names a comma is replaced
+  // by _$_ and a closing paranthesis is replaced by _$ and opening parathesis
+  // is replaced by $_ since commas can appear only in middle of two type names
+  // in mapping, 't' is guaranteed to come after _$_ in yul name Its important
+  // to handle this case because we can have two closing parathesis in the
+  // nested mapping case so we we have t_mapping(xxx, t_mapping(yyy,zzz))
+  // which has the yul type name t_mapping$_xxx_$_t_mapping$_yyy_$_zzz_$_$
+  // Here the second last closing parathesis is mis interpreted as comma.
```
Comment (Technius) [inline lib/libYulAST/Types/YulTypeContext.cpp:692]: Can we format this a bit more nicely? This is hard to read.

File: lib/libyul2llvm/YulASTVisitor/IntrinsicEmitter.cpp (hunk 1 above)
Comment (veriAjinkya) [reply lib/libyul2llvm/YulASTVisitor/IntrinsicEmitter.cpp:576]: done

File: lib/libyul2llvm/YulASTVisitor/IntrinsicRewriter.cpp (hunk 2 above)
Comment (veriAjinkya) [reply lib/libyul2llvm/YulASTVisitor/IntrinsicRewriter.cpp:465]: Its part of the debug message, printing out to err stream instead

File: lib/libYulAST/Types/YulTypeContext.cpp (hunk 3 above)
Comment (veriAjinkya) [reply lib/libYulAST/Types/YulTypeContext.cpp:692]: done

**Review by Technius — APPROVED** [review]

**Pull request merged**